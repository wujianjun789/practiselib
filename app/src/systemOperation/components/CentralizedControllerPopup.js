import React, {Component} from 'react';
import Panel from '../../components/Panel';
import PanelFooter from '../../components/PanelFooter';
import Select from '../../components/Select.1';
import MapView from '../../components/MapView';
import PropTypes from 'prop-types';

import {ObjectPerValid, Name2Valid, latlngValid, MACValid} from '../../util/index'
export default class CentralizedControllerPopup extends Component {
    constructor(props) {
        super(props);
        const {id="", name="", modelId="", model="", domainId="", domain="", lng=0, lat=0} = props.data;
        this.state = {
            id:id,
            name:name,
            model:model,
            modelId:modelId,
            domain:domain,
            domainId:domainId,
            lng:lng,
            lat:lat
        };

        this.onChange = this.onChange.bind(this);
        this.onCancel = this.onCancel.bind(this);
        this.onConfirm = this.onConfirm.bind(this);

        this.mapDragend = this.mapDragend.bind(this);
    }

    onChange(e) {
        let id = e.target.id;
        if (id == "model") {
            this.setState({modelId: this.props.modelList.options[e.target.selectedIndex].id});
        }

        if (id == "domain") {
            this.setState({modelId: this.props.domainList.options[e.target.selectedIndex].id});
        }

        let value = e.target.value;
        let newValue='';
        if(id == "lat" || id == "lng"){
            newValue = Number(ObjectPerValid(value, latlngValid));
        }else if(id == "name"){
            newValue = ObjectPerValid(value, Name2Valid);
        }else if(id == "id") {
            newValue = ObjectPerValid(value, MACValid)
        }
        else{
            newValue = value;
        }

        this.setState({[id]: newValue});
    }

    onCancel() {
        this.props.overlayerHide();
    }

    onConfirm() {
        this.props.onConfirm && this.props.onConfirm(this.state);
        // this.props.overlayerHide();
    }

    mapDragend(data) {
        for(let key in data.latlng){
            let value = data.latlng[key];
            let newValue = Number(ObjectPerValid(""+value, latlngValid), latlngValid);
            this.setState({[key]:newValue});
        }
    }

    render() {
        const {className, title, domainList, modelList, popId} = this.props;
        const {id, name, model, domain, lng, lat} = this.state;
        const footer = <PanelFooter funcNames={['onCancel','onConfirm']} btnTitles={['取消','确认']}
                                  btnClassName={['btn-default', 'btn-primary']} btnDisabled={[false, false]}
                                  onCancel={this.onCancel} onConfirm={this.onConfirm}/>;
        return (
            <div className={className}>
                <Panel title={title} closeBtn={true} closeClick={this.onCancel}>
                    <div className="popup-left">
                        <div className="form-group clearfix">
                            <label htmlFor="id" className="col-sm-4 control-label">设备编号：</label>
                            <div className="col-sm-8">
                                <input type="text" className="form-control" id="id" placeholder="id" value={id}
                                       onChange={this.onChange} disabled={popId=='edit'?true:false}/>
                            </div>
                        </div>
                        <div className="form-group clearfix">
                            <label htmlFor="name" className="col-sm-4 control-label">名称：</label>
                            <div className="col-sm-8">
                                <input type="text" className="form-control" id="name" placeholder="name" value={name}
                                       onChange={this.onChange}/>
                            </div>
                        </div>
                        <div className="form-group clearfix">
                            <label htmlFor="model" className="col-sm-4 control-label">型号：</label>
                            <div className="col-sm-8">
                                <Select id="model" className="form-control" titleField={modelList.titleField}
                                        valueField={modelList.valueField} options={modelList.options} value={model}
                                        onChange={this.onChange}/>
                            </div>
                        </div>
                        <div className="form-group clearfix">
                            <label htmlFor="domain" className="col-sm-4 control-label">域：</label>
                            <div className="col-sm-8">
                                <Select id="domain" className="form-control" titleField={domainList.titleField}
                                        valueField={domainList.valueField} options={domainList.options} value={domain}
                                        onChange={this.onChange}/>
                            </div>
                        </div>
                        <div className="form-group clearfix">
                            <label htmlFor="lng" className="col-sm-4 control-label">经度：</label>
                            <div className="col-sm-8">
                                <input type="text" className="form-control" id="lng" placeholder="lng" value={lng}
                                       onChange={this.onChange}/>
                            </div>
                        </div>
                        <div className="form-group clearfix">
                            <label htmlFor="lat" className="col-sm-4 control-label">纬度：</label>
                            <div className="col-sm-8">
                                <input type="text" className="form-control" id="lat" placeholder="lat" value={lat}
                                       onChange={this.onChange}/>
                            </div>
                        </div>
                        {footer}
                    </div>
                    <div className="popup-map">
                        <MapView option={{mapZoom: false }} mapData={{id: "CentralizedPopup", latlng:{lng:lng, lat:lat},
                            position: [ {"device_id": id, "device_type": "DEVICE", lng: lng, lat: lat } ], data: [ { id: id, name: name } ] } }
                                 mapCallFun={{mapDragendHandler:this.mapDragend}} />
                    </div>
                </Panel>
            </div>
        )
    }
}

CentralizedControllerPopup.propTypes = {
    popId: PropTypes.string.isRequired
}
