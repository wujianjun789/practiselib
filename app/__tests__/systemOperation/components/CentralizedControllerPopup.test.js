import React, {Component} from 'react';
import {shallow, mount} from 'enzyme';
import renderer from 'react-test-renderer';
import CentralizedControllerPopup from '../../../src/systemOperation/components/CentralizedControllerPopup';

describe("<CentralizedControllerPopup />", () => {
    const popId = 'add';
    const title = '添加设备';
    const className='centralized-popup';
    const modelList = {
        titleField: 'title',
        valueField: 'value',
        options: [
            {id: 1, title: 'model01', value: 'model01'},
            {id: 2, title: 'model02', value: 'model02'},
            {id: 3, title: 'model03', value: 'model03'},
            {id: 4, title: 'model04', value: 'model04'},
            {id: 5, title: 'model05', value: 'model05'},
            {id: 6, title: 'model06', value: 'model06'},
            {id: 7, title: 'model07', value: 'model07'}
        ]
    };
    const domainList = {
        titleField: 'title',
        valueField: 'value',
        options: [
            {id: 1, title: 'domain01', value: 'domain01'},
            {id: 2, title: 'domain02', value: 'domain02'},
            {id: 3, title: 'domain03', value: 'domain03'},
            {id: 4, title: 'domain04', value: 'domain04'},
            {id: 5, title: 'domain05', value: 'domain05'},
            {id: 6, title: 'domain06', value: 'domain06'},
            {id: 7, title: 'domain07', value: 'domain07'}
        ]
    };
    const data = {
        id: 'sdf',
        name: 'name01',
        model: 'model01',
        domain: 'domain01',
        lng: 121.49971691534425,
        lat: 31.239658843127756
    };
    it('render normal', () => {
        const cmp = shallow(<CentralizedControllerPopup popId={popId} className={className} title={title} data={data} domainList={domainList} modelList={modelList}/>);
        const container = cmp.find(`.${className}`);
        expect(container.length).toBe(1);
        
        const panel = cmp.find("Panel");
        expect(panel.length).toBe(1);
        expect(panel.prop('title')).toBe(title);

        const items = cmp.find('.form-group.clearfix');
        expect(items.length).toBe(6);

        const name = items.at(1);
        const lab = name.find('.control-label');
        expect(lab.text()).toBe('名称：');
        const input = name.find('#name');
        expect(input.props().value).toBe(data.name);

        expect(items.at(0).props().disabled).not.toBeTruthy();

        const domainSelect = items.at(3).find('Select');
        expect(domainSelect.prop('titleField')).toBe(domainList.titleField);
        expect(domainSelect.prop('valueField')).toBe(domainList.valueField);
        expect(domainSelect.props().options).toBe(domainList.options);
        expect(domainSelect.props().value).toBe(data.domain);
    });

    it('render with popId="edit" to disabled input#id', () => {
        const cmp = shallow(<CentralizedControllerPopup popId='edit' className={className} title={title} data={data} domainList={domainList} modelList={modelList}/>);
        expect(cmp.find('#id').props().disabled).toBeTruthy();
    });

    it('simulate event', () => {
        const cmp = shallow(<CentralizedControllerPopup popId='edit' className={className} title={title} data={data} domainList={domainList} modelList={modelList}/>);

        let name = cmp.find('#name');
        expect(name.prop('value')).toBe(data.name);
        name.simulate('change', {target: {value: 'name02', id: 'name'}});
        name = cmp.find('#name');
        expect(name.props().value).toBe('name02');

        let modelSelect = cmp.find('#model');
        expect(modelSelect.prop('value')).toBe(data.model);
        modelSelect.simulate('change', {target: {value: 'model02', id: 'model'}});
        modelSelect = cmp.find('#model');
        expect(modelSelect.props().value).toBe('model02');
    });

    it('snapshot with popId="add"', () => {
        const cmp = renderer.create(<CentralizedControllerPopup popId='add' className={className} title={title} data={data} domainList={domainList} modelList={modelList}/>);
        const tree = cmp.toJSON();
        expect(tree).toMatchSnapshot();
    });
    
    it('snapshot with popId="edit"', () => {
        const cmp = renderer.create(<CentralizedControllerPopup popId='edit' className={className} title={title} data={data} domainList={domainList} modelList={modelList}/>);
        const tree = cmp.toJSON();
        expect(tree).toMatchSnapshot();
    });
})