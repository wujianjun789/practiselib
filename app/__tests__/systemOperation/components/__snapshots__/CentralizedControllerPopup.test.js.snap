exports[`<CentralizedControllerPopup /> snapshot with popId="add" 1`] = `
<div
  className="centralized-popup">
  <div
    className="modal "
    role="dialog">
    <div
      className="modal-dialog"
      role="document">
      <div
        className="modal-content">
        <div
          className="modal-header">
          <button
            aria-label="Close"
            className="close"
            data-dismiss="modal"
            onClick={[Function]}
            type="button">
            <span
              aria-hidden="true">
              ×
            </span>
          </button>
          <h4
            className="modal-title">
            添加设备
          </h4>
        </div>
        <div
          className="modal-body">
          <div
            className="popup-left">
            <div
              className="form-group clearfix">
              <label
                className="col-sm-4 control-label"
                htmlFor="id">
                设备编号：
              </label>
              <div
                className="col-sm-8">
                <input
                  className="form-control"
                  disabled={false}
                  id="id"
                  onChange={[Function]}
                  placeholder="id"
                  type="text"
                  value="sdf" />
              </div>
            </div>
            <div
              className="form-group clearfix">
              <label
                className="col-sm-4 control-label"
                htmlFor="name">
                名称：
              </label>
              <div
                className="col-sm-8">
                <input
                  className="form-control"
                  id="name"
                  onChange={[Function]}
                  placeholder="name"
                  type="text"
                  value="name01" />
              </div>
            </div>
            <div
              className="form-group clearfix">
              <label
                className="col-sm-4 control-label"
                htmlFor="model">
                型号：
              </label>
              <div
                className="col-sm-8">
                <select
                  className="select form-control"
                  id="model"
                  onChange={[Function]}
                  value="model01">
                  <option
                    value="model01">
                    model01
                  </option>
                  <option
                    value="model02">
                    model02
                  </option>
                  <option
                    value="model03">
                    model03
                  </option>
                  <option
                    value="model04">
                    model04
                  </option>
                  <option
                    value="model05">
                    model05
                  </option>
                  <option
                    value="model06">
                    model06
                  </option>
                  <option
                    value="model07">
                    model07
                  </option>
                </select>
              </div>
            </div>
            <div
              className="form-group clearfix">
              <label
                className="col-sm-4 control-label"
                htmlFor="domain">
                域：
              </label>
              <div
                className="col-sm-8">
                <select
                  className="select form-control"
                  id="domain"
                  onChange={[Function]}
                  value="domain01">
                  <option
                    value="domain01">
                    domain01
                  </option>
                  <option
                    value="domain02">
                    domain02
                  </option>
                  <option
                    value="domain03">
                    domain03
                  </option>
                  <option
                    value="domain04">
                    domain04
                  </option>
                  <option
                    value="domain05">
                    domain05
                  </option>
                  <option
                    value="domain06">
                    domain06
                  </option>
                  <option
                    value="domain07">
                    domain07
                  </option>
                </select>
              </div>
            </div>
            <div
              className="form-group clearfix">
              <label
                className="col-sm-4 control-label"
                htmlFor="lng">
                经度：
              </label>
              <div
                className="col-sm-8">
                <input
                  className="form-control"
                  id="lng"
                  onChange={[Function]}
                  placeholder="lng"
                  type="text"
                  value={121.49971691534425} />
              </div>
            </div>
            <div
              className="form-group clearfix">
              <label
                className="col-sm-4 control-label"
                htmlFor="lat">
                纬度：
              </label>
              <div
                className="col-sm-8">
                <input
                  className="form-control"
                  id="lat"
                  onChange={[Function]}
                  placeholder="lat"
                  type="text"
                  value={31.239658843127756} />
              </div>
            </div>
            <div
              className="modal-footer">
              <button
                className="btn btn-default"
                disabled={false}
                onClick={[Function]}
                type="button">
                取消
              </button>
              <button
                className="btn btn-primary"
                disabled={false}
                onClick={[Function]}
                type="button">
                确认
              </button>
            </div>
          </div>
          <div
            className="popup-map">
            <div
              className="map-view "
              id="CentralizedPopup" />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<CentralizedControllerPopup /> snapshot with popId="edit" 1`] = `
<div
  className="centralized-popup">
  <div
    className="modal "
    role="dialog">
    <div
      className="modal-dialog"
      role="document">
      <div
        className="modal-content">
        <div
          className="modal-header">
          <button
            aria-label="Close"
            className="close"
            data-dismiss="modal"
            onClick={[Function]}
            type="button">
            <span
              aria-hidden="true">
              ×
            </span>
          </button>
          <h4
            className="modal-title">
            添加设备
          </h4>
        </div>
        <div
          className="modal-body">
          <div
            className="popup-left">
            <div
              className="form-group clearfix">
              <label
                className="col-sm-4 control-label"
                htmlFor="id">
                设备编号：
              </label>
              <div
                className="col-sm-8">
                <input
                  className="form-control"
                  disabled={true}
                  id="id"
                  onChange={[Function]}
                  placeholder="id"
                  type="text"
                  value="sdf" />
              </div>
            </div>
            <div
              className="form-group clearfix">
              <label
                className="col-sm-4 control-label"
                htmlFor="name">
                名称：
              </label>
              <div
                className="col-sm-8">
                <input
                  className="form-control"
                  id="name"
                  onChange={[Function]}
                  placeholder="name"
                  type="text"
                  value="name01" />
              </div>
            </div>
            <div
              className="form-group clearfix">
              <label
                className="col-sm-4 control-label"
                htmlFor="model">
                型号：
              </label>
              <div
                className="col-sm-8">
                <select
                  className="select form-control"
                  id="model"
                  onChange={[Function]}
                  value="model01">
                  <option
                    value="model01">
                    model01
                  </option>
                  <option
                    value="model02">
                    model02
                  </option>
                  <option
                    value="model03">
                    model03
                  </option>
                  <option
                    value="model04">
                    model04
                  </option>
                  <option
                    value="model05">
                    model05
                  </option>
                  <option
                    value="model06">
                    model06
                  </option>
                  <option
                    value="model07">
                    model07
                  </option>
                </select>
              </div>
            </div>
            <div
              className="form-group clearfix">
              <label
                className="col-sm-4 control-label"
                htmlFor="domain">
                域：
              </label>
              <div
                className="col-sm-8">
                <select
                  className="select form-control"
                  id="domain"
                  onChange={[Function]}
                  value="domain01">
                  <option
                    value="domain01">
                    domain01
                  </option>
                  <option
                    value="domain02">
                    domain02
                  </option>
                  <option
                    value="domain03">
                    domain03
                  </option>
                  <option
                    value="domain04">
                    domain04
                  </option>
                  <option
                    value="domain05">
                    domain05
                  </option>
                  <option
                    value="domain06">
                    domain06
                  </option>
                  <option
                    value="domain07">
                    domain07
                  </option>
                </select>
              </div>
            </div>
            <div
              className="form-group clearfix">
              <label
                className="col-sm-4 control-label"
                htmlFor="lng">
                经度：
              </label>
              <div
                className="col-sm-8">
                <input
                  className="form-control"
                  id="lng"
                  onChange={[Function]}
                  placeholder="lng"
                  type="text"
                  value={121.49971691534425} />
              </div>
            </div>
            <div
              className="form-group clearfix">
              <label
                className="col-sm-4 control-label"
                htmlFor="lat">
                纬度：
              </label>
              <div
                className="col-sm-8">
                <input
                  className="form-control"
                  id="lat"
                  onChange={[Function]}
                  placeholder="lat"
                  type="text"
                  value={31.239658843127756} />
              </div>
            </div>
            <div
              className="modal-footer">
              <button
                className="btn btn-default"
                disabled={false}
                onClick={[Function]}
                type="button">
                取消
              </button>
              <button
                className="btn btn-primary"
                disabled={false}
                onClick={[Function]}
                type="button">
                确认
              </button>
            </div>
          </div>
          <div
            className="popup-map">
            <div
              className="map-view "
              id="CentralizedPopup" />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
